tests/user/test_user_auth.py

```diff
@@ -0,0 +1,151 @@
+import base64
+from datetime import datetime, timedelta
+import json
+import jwt
+import pytest
+from unittest.mock import Mock, patch
+from typing import Generator, Union
+from personal_finances.user.user_auth import user_handler, create_user_hash_password
+
+
+TEST_USER_ID = "testuser"
+TEST_USER_PASSWORD = "123"
+TEST_USER_DYNAMO_RESPONSE = {
+    "Item": {
+        "userId": {"S": TEST_USER_ID},
+        "password": {
+            # password "123" hashed by bcrypt
+            "S": "$2b$12$735jLXvW38meOqPXIwhsSO1bNP6AcFY7K4K3WuPiDsdD7wajOrhp2"
+        },
+    }
+}
+TEST_JWT_SECRETS = "jwt_secret"
+TEST_DATE_TIME_NOW = datetime(2060, 1, 1, 10, 00, 00)
+TEST_ENVAR_DICT = {
+    "INFINEAT_DYNAMODB_USER_TABLE_NAME": "table_name",
+    "INFINEAT_JWT_SECRET_ID": "jwt_secret",
+}
+
+
+def _encode_basic_auth(user: str, password: str) -> str:
+    credentials = f"{user}:{password}"
+    credentials_encoded = credentials.encode()
+    authorization_encoded = base64.b64encode(credentials_encoded)
+    authorization = f"Basic {authorization_encoded.decode()}"
+    return authorization
+
+
+@pytest.fixture(autouse=True)
+def os_mock() -> Generator[Mock, None, None]:
+    with patch("personal_finances.user.user_auth.os") as mock:
+        yield mock
+
+
+@pytest.fixture(autouse=True)
+def boto3_mock() -> Generator[Mock, None, None]:
+    with patch("personal_finances.user.user_auth.boto3.client") as mock:
+        yield mock
+
+
+@pytest.fixture(autouse=True)
+def datetime_mock() -> Generator[Mock, None, None]:
+    with patch("personal_finances.user.user_auth.datetime") as mock:
+        mock.now.return_value = TEST_DATE_TIME_NOW
+        yield mock
+
+
+def _generate_test_jwt_token() -> str:
+    return jwt.encode(
+        {"exp": TEST_DATE_TIME_NOW + timedelta(hours=1), "userId": TEST_USER_ID},
+        TEST_JWT_SECRETS,
+        algorithm="HS256",
+    )
+
+
+@pytest.mark.parametrize(
+    "authorization_header, expected_response, dynamo_response_dict, env_var_dict",
+    [
+        (
+            _encode_basic_auth(TEST_USER_ID, "incorrect_password"),
+            {"statusCode": 401, "body": "User or password incorrect"},
+            TEST_USER_DYNAMO_RESPONSE,
+            TEST_ENVAR_DICT,
+        ),
+        (
+            _encode_basic_auth("incorrect_user", "incorrect_password"),
+            {"statusCode": 401, "body": "User or password incorrect"},
+            {"usernot": "found"},
```

Comment:
This is an invalid dynamodb response, dynamodb always answers it's own typed dictionary. That's a good test case, but it's a case where we should return `5xx` to the user, because the backend database is misbehaving.

In your tests and main code you can use the `from boto3.dynamodb.types import TypeDeserializer, TypeSerializer`

Have a look a this link, https://aws.amazon.com/blogs/database/exploring-amazon-dynamodb-sdk-clients/